### ISSUE TITLE
Question about training UNet with the provided generated images

### ISSUE THREAD

**Issue by DaiZhewei** [sjtuplayer/anomalydiffusion#38]
Hi Author,
We've trained 15 UNets using the provided generated data and conducted segmentation tests. We would like to ask you some questions:
1. The provided data only contains 500 image-mask pairs per defect class, which contradicts the 1000 pairs mentioned in the paper.
2. The resolution of the provided data is 256, while the stable-diffusion model directly generates images with a resolution of 512. How were the 256 resolution images obtained?
3. The segmentation results of the UNet we trained using the provided data on the test set are significantly different from those reported in the paper. Could you please provide insight into what might be causing this?
【Results in the paper】
Image Level
AUROC: 99.2 AP: 99.7 F1: 98.7
Pixel Level
AUROC: 99.1 AP: 81.4 F1: 76.3
【Results we implemented】
Image Level
AUROC: 94.92 AP: 97.54 F1: 93.83
Pixel Level
AUROC: 95.46 AP: 64.31 F1: 64.21
Looking forward to your reply.

**Comment by DaiZhewei** [comment]
> Dear Dai Zhewei,
> 
> Stable diffusion model can utilize any resolution which is multiple of 64. So it can be higher or lower. I am not sure about UNet, but they trained classifier on the provided dataset.

OK, I undestand, thanks for your reply!

**Comment by engrmusawarali** [comment]
Dear @DaiZhewei  did you get the segmentation results for all classes of MVTec and computed average?

**Comment by DaiZhewei** [comment]
> Dear @DaiZhewei did you get the segmentation results for all classes of MVTec and computed average?

Yes, this is the average of the results across all classes.

**Comment by sjtuplayer** [comment]
> Hi Author, We've trained 15 UNets using the provided generated data and conducted segmentation tests. We would like to ask you some questions:
> 
> 1. The provided data only contains 500 image-mask pairs per defect class, which contradicts the 1000 pairs mentioned in the paper.
> 2. The resolution of the provided data is 256, while the stable-diffusion model directly generates images with a resolution of 512. How were the 256 resolution images obtained?
> 3. The segmentation results of the UNet we trained using the provided data on the test set are significantly different from those reported in the paper. Could you please provide insight into what might be causing this?
>    【Results in the paper】
>    Image Level
>    AUROC: 99.2 AP: 99.7 F1: 98.7
>    Pixel Level
>    AUROC: 99.1 AP: 81.4 F1: 76.3
>    【Results we implemented】
>    Image Level
>    AUROC: 94.92 AP: 97.54 F1: 93.83
>    Pixel Level
>    AUROC: 95.46 AP: 64.31 F1: 64.21
>    Looking forward to your reply.

1. We abandoned 500 image-pairs with lower quality as illustrated in the reamdme. If you need the other 500 images, you can use the offered pretrained model to generate them
2. The data is downsampled into 256 size for the downstream anomaly detection tasks.
3. The AP score should not be so bad, if you use both the anomaly detection code and generated data, it should be much higher, you can try it again, since we have already checked the performance when running the open-sourced code.

**Comment by CVKim** [comment]
@DaiZhewei 

Hello,

I was wondering if the code for the U-Net model for the 15 classes of MVTEC-AD mentioned above is provided in the AnomalyDiffusion GitHub repository? If it is available, could you please share it with me as well? I would like to attempt a performance evaluation. I would really appreciate your help!!!

**Comment by CVKim** [comment]
@sjtuplayer 

Hello,

Does running the open-source code mean that you used the U-Net segmentation code from the Anomaly Diffusion official GitHub?

I am a student learning about diffusion, and I would like to reproduce the segmentation results for the Anomaly Diffusion model. Could you please explain the detailed evaluation methods?

For example, did you train models individually for each class in the MVTEC-AD dataset, or did you integrate the defect types under a class (like hazelnut) into a single model for training?

For the train/test set preparation, did you use the images generated by the Anomaly Diffusion model, or did you also include the actual MVTEC-AD dataset?

I apologize for the many questions, but your assistance would be greatly appreciated!!

**Comment by DaiZhewei** [comment]
> @DaiZhewei
> 
> Hello,
> 
> I was wondering if the code for the U-Net model for the 15 classes of MVTEC-AD mentioned above is provided in the AnomalyDiffusion GitHub repository? If it is available, could you please share it with me as well? I would like to attempt a performance evaluation. I would really appreciate your help!!!

Hello, for training the UNet, you can refer to "anomalydiffusion/train-localization.py" and "anomalydiffusion/unet_utils/".

**Comment by CVKim** [comment]
@DaiZhewei 

Thank you!

I couldn't find the details in the paper, so I have a few questions. I would also like to train a segmentation model using U-Net and evaluate its performance.

Was the segmentation done using only the approximately 500 generated images that were provided?
When training the segmentation model, what split ratio was used for the train/validation sets?

Additionally, should I understand that you have implemented a separate code for segmentation training using the scripts provided in the "unet_utils" folder?

I apologize for these seemingly minor questions... I'm a university student who wants to learn about deep learning and I'm trying to explore and experiment with these concepts on my own.

【Results in the paper】
Image Level
AUROC: 99.2 AP: 99.7 F1: 98.7
Pixel Level
AUROC: 99.1 AP: 81.4 F1: 76.3
【Results we implemented】
Image Level
AUROC: 94.92 AP: 97.54 F1: 93.83
Pixel Level
AUROC: 95.46 AP: 64.31 F1: 64.21

Looking forward to your reply!!

**Comment by DaiZhewei** [comment]
@CVKim 
Hello,
1. Yes, I used the 500 anomalous image-mask pairs provided, which is recommended by the author in this issue. "We abandoned 500 image-pairs with lower quality as illustrated in the reamdme. If you need the other 500 images, you can use the offered pretrained model to generate them."
2. The training sets are the 500 anomalous image-mask pairs and the training sets in MVTec, and the validation set is the test set in the paper ("reserving the remaining two-thirds for testing"). For more details, you can refer to the "data_loader.py" file in the "unet_utils" folder. The "class MVTec_Anomaly_Detection(Dataset)" is for training set, and the "class MVTecDRAEMTestDataset_partial(Dataset)" is for validation set. 
3. No, I followed the official implementation, not a separate code. The details are provided in the "unet_utils" folder.
